### DISCUSSION TITLE
navigate method vs push method

### DISCUSSION THREAD

**Discussion by ferrarafer** [Milad-Akarie/auto_route_library#1498] [Q&A] [answered]
Hello @Milad-Akarie , I wondering which are the differences between `navigate` and `push` methods. From the documentation I understand that push always add an entry to the stack and navigate pops the routes on the stack if the route already exists on the stack. Also said, navigating without context is not recommended in nested navigation unless you use navigate instead of push but it's not clear which are the problems of using push.

**Context**
We are using AutoRoute in a project but in generally we use navigate method and we want to know if can be replaced by push method without side effects. We want to do the replacement because navigate method does not await for the call and push yes, so navigate is **not** working when we want to pass a value back from another route.

**Comment by Milad-Akarie** [comment, marked as answer]
@ferrarafer I always seem to miss the threads here because the stupid **Discussions** tap has no counter badge.

Well  "navigating without context is not recommended in nested navigation unless you use navigate instead of push" is only applicable if you work with absolute paths, but I guess it's  relatively old now, in recent versions auto_route can auto find the route-scope and push it to the right router.    


if you have a stack that looks like the following: 
`/products/12`

and you **push**  `/products/12/reviews` that's a whole new stack added to the hierarchy,

but if you **navigate** to  `/products/12/reviews `only `reviews` will be added to already existing stack

and if you only push 'reviews' then  you'll get the same result as using the navigate method

but why would I need to use absolute paths you might think? 
giving the following routes setup

```yaml
[
  /venders
        :id/reviews
        :id/products
             :id/reviews
]
```
let's we're on  `/venders/2/products`
from here calling `rootRouter.push('id/reviews')` will get us the first match from the top of the hierarchy and that's `products/id/reviews` when we meant to navigate to `/venders/id/reviews` , in such case you'd need to provide full path,
`ootRouter.navigateTo('/venders/id/reviews')`

but of course this does not apply to `PageRouteInfos` with different names. 


so I guess you can safely  switch to push if you're using  unique `PageRouteInfos` to navigate
